airflow/utils/state.py

```diff
@@ -99,6 +99,7 @@ def finished(cls):
             cls.SUCCESS,
             cls.FAILED,
             cls.SKIPPED,
+            cls.UPSTREAM_FAILED
```

Comment:
This is wrong. A task in UPSTREAM_FAILED has never been attempted.

Reply:
The doc for this list says:

> A list of states indicating that a task started and completed a
> run attempt.

Tasks in upstream_failed fail at the first hurdle: they have never been started.

Regardless of wether this change is right in the long run if falls way outside the title of the PR of "collect trigger rule dep check per dag run" to a behaviour change.

Reply:
I absolutely agree with what Ash is saying. The documentation doesn’t match the implementation and the change would probably need a separate ticket, because it changes the basic concept or meaning of finished tasks.

Reply:
Great, I will change it